File: BaekJoon_java/Main_5212.java
import java.util.Scanner;

public class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int r = sc.nextInt();
		int c = sc.nextInt();

		char[][] map = new char[r + 2][c + 2];
		for (int i = 0; i < r+2; i++)
			for (int j = 0; j < c+2; j++)
				map[i][j] = '.';

		for (int i = 1; i < r + 1; i++) {
			String s = sc.next();
			for (int j = 1; j < c + 1; j++)
				map[i][j] = s.charAt(j - 1);
		}
		
		char[][] newmap = new char[r + 2][c + 2];
		for (int i = 0; i < r + 2; i++) {
			for (int j = 0; j < c + 2; j++)
				newmap[i][j] = map[i][j];
		}
		

		for (int i = 1; i < r + 1; i++) {
			for (int j = 1; j < c + 1; j++) {
				int count = 0;
				if (map[i][j] == 'X') {
					if (map[i - 1][j] == '.')
						count++;
					if (map[i][j - 1] == '.')
						count++;
					if (map[i + 1][j] == '.')
						count++;
					if (map[i][j + 1] == '.')
						count++;
					if (count >= 3)
						newmap[i][j] = '.';
				}
			}
		}
		
//		for (int i = 0; i < r+2; i++) {
//			for (int j = 0; j < c+2; j++)
//				System.out.print(newmap[i][j]);
//			System.out.println();
//		}

		int startr = r + 1;
		int lastr = 0;
		int startc = c + 1;
		int lastc = 0;

		for (int i = 0; i < r + 2; i++) {
			for (int j = 0; j < c + 2; j++) {
				if (newmap[i][j] == 'X') {
					startr = (i < startr) ? i : startr;
					startc = (j < startc) ? j : startc;
					lastr = (i > lastr) ? i : lastr;
					lastc = (j > lastc) ? j : lastc;
				}
			}
		}
		
//		System.out.println (startr + " " + lastr+ " " + startc + " " + lastc);

//		for (int i = startr; i <= lastr; i++) {
//			for (int j = startc; j <= lastc; j++)
//				System.out.print(newmap[i][j]);
//			System.out.println();
//		}
		
		for (int i = startr ; i <= lastr; i++) {
			for (int j = startc; j <= lastc; j++)
				System.out.print(newmap[i][j]);
			System.out.println();
		}
		sc.close();
	}
}

File: BaekJoon_c++/Main_5212.cpp
#include <iostream>
using namespace std;

int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(NULL);
	cout.tie(NULL);

	int R, C;
	char map[12][12]; // 지도로 출력되지 않는 곳도 탐색해봐야함.
	cin >> R >> C;

	// 지도 초기화
	for (int r = 0; r < R+2; r++) {
		for (int c = 0; c < C+2; c++) {
			map[r][c] = '.';
		}
	}

	// 지도 입력 받기 ( 시작하는 인덱스 주의 ! )
	for (int r = 1; r < R + 1; r++) {
		for (int c = 1; c < C + 1; c++) {
			char ch;
			cin >> ch;

			map[r][c] = ch;
		}
	}

	// 지도 복사
	char newmap[12][12];
	for (int r = 0; r < R + 2; r++) {
		for (int c = 0; c < C + 2; c++)
			newmap[r][c] = map[r][c];
	}

	// 지도 탐색하기
	for (int r = 1; r < R + 1; r++) {
		for (int c = 1; c < C + 1; c++) {
			int count = 0; // 인접한 면 중 바다를 세기

			if (map[r][c] == 'X') {
				if (map[r - 1][c] == '.') // 상
					count++;
				if (map[r][c- 1] == '.') // 좌
					count++;
				if (map[r + 1][c] == '.') // 하
					count++;
				if (map[r][c + 1] == '.') // 우
					count++;
				if (count >= 3)
					newmap[r][c] = '.';
			}
		}
	}

	// 최소 지도 범위 구하기
	int startr = R + 1;
	int lastr = 0;
	int startc = C + 1;
	int lastc = 0;

	for (int r = 0; r < R + 2; r++) {
		for (int c = 0; c < C + 2; c++) {
			if (newmap[r][c] == 'X') {
				startr = (r < startr) ? r : startr;
				startc = (c < startc) ? c : startc;
				lastr = (r > lastr) ? r: lastr;
				lastc = (c > lastc) ? c : lastc;
			}
		}
	}
	

	for (int r = startr; r <= lastr; r++) {
		for (int c = startc; c <= lastc; c++)
			cout << newmap[r][c];
		cout << endl;
	}
	



}